server/member/lease.go

```diff
@@ -162,51 +162,51 @@ func (l *lease) getExpireTime() time.Time {
 
 // `renewLeaseBg` keeps the lease alive by periodically call `lease.KeepAliveOnce`.
 // The l.expireTime will be updated during renewing and the renew lease result (whether alive) will be told to caller by `renewed` channel.
-func (l *lease) renewLeaseBg(ctx context.Context, interval time.Duration, renewed chan<- bool) {
-	l.logger.Info("start renewing lease background", zap.Duration("interval", interval))
-	defer l.logger.Info("stop renewing lease background", zap.Duration("interval", interval))
+func (l *lease) renewLeaseBg(ctx context.Context, renewed chan<- bool) {
+	l.logger.Info("start renewing lease background", zap.Int64("lease-id", int64(l.ID)))
+	defer l.logger.Info("stop renewing lease background", zap.Int64("lease-id", int64(l.ID)))
+
+	keepaliveStream, err := l.rawLease.KeepAlive(ctx, l.ID)
+	if err != nil {
+		l.logger.Error("fail to start keep lease alive stream", zap.Int64("lease-id", int64(l.ID)), zap.Error(err))
+		return
+	}
 
-L:
 	for {
-		renewOnce := func() renewLeaseResult {
-			start := time.Now()
-			ctx1, cancel := context.WithTimeout(ctx, l.timeout)
-			defer cancel()
-			resp, err := l.rawLease.KeepAliveOnce(ctx1, l.ID)
-			if err != nil {
-				l.logger.Error("lease keep alive failed", zap.Error(err))
-				return renewLeaseFailed
-			}
-			if resp.TTL < 0 {
-				l.logger.Warn("lease is expired")
-				return renewLeaseExpired
+		start := time.Now()
+		select {
+		case resp, ok := <-keepaliveStream:
```

Comment:
What will happens if the `keepaAliveSteam` is disconnected?  In the previous implementation, if the retry succeeds before the lease timeout, a single `keepAliveOnce` failure will not lead to a new election, which has a certain degree of fault tolerance. But in the current implementation, once steam is disconnected, there will be no further attempts to renew the lease. Maybe we need try to reconnected when `keepaAliveSteam` is disconnected.

Reply:
Make senses. It should retry if the stream is broken.